### ISSUE TITLE
Skipped steps showing as failed in Report Portal UI

### ISSUE THREAD

**Issue by Rickdickulous** [reportportal/agent-Python-RobotFramework#145]
**Describe the bug**
Steps that are skipped in Robot Framework test execution (IF/ELSE trees) sometimes show as failed in Report Portal UI.

Please let me know if more context is needed. I can't find any particular patterns to make it happen as it seems random.
The same tests pass (in Report Portal) 95% of the time, but now and then at random we see this issue.

Below is a screenshot of the steps that are being run and the log output from Report Portal. All these steps passed in the Robot Framework log but are showing as fail in Report Portal. Let me know if you need part of the log or xml output or anything. I can't share full logs so please be specific and I'll get you what you need. Alternatively I can try to recreate the occurrence on sample tests that I can share. Thank you for your help.

![bug_example](https://user-images.githubusercontent.com/22482646/174607256-71a745b4-0412-497d-b1ff-191dc11f35f9.png)

**Steps to Reproduce**
1. Run a Robot Framework test with an IF statement that is not executed, with the Report Portal Robot Framework listener called to send results to Report Portal.
2. Review results in Report Portal and see the test is marked as failed, because the skipped step is marked as failed in the log view.

**Expected behavior**
Tests that pass, even with skipped steps, are marked as passed.

**Actual behavior**
Steps that are skipped are sometimes being marked as failed in Report Portal.

**Package versions**
reportportal-client 5.7.0
 robotframework-reportportal 5.2.1

**Comment by HardNorth** [comment]
Unfortunately my sample test case gives me correct result:
```robot
*** Settings ***
Documentation  Example of skipping falure startement by 'IF'

*** Variables ***
${SKIP}    True

*** Test Cases ***
Test If Skip
    Log  One passed step to check
    IF   not ${SKIP}
         Fail  This step should be skipped
    END

```
![Screen Shot 2022-06-20 at 5 44 07 PM](https://user-images.githubusercontent.com/4364466/174627028-45c34ff6-9139-4608-bfc4-c54281652c53.png)

I tried this case several times and the result is the same. Do you have any traceback in your logs?

**Comment by Rickdickulous** [comment]
I have another round of info for you.

I setup a Robot Framework test to run the if statement 1000 times per run. I ran this around 10 times and saw the bug 5 times.

### Observations
1. I noticed that sometimes the results in Report Portal would first show correctly that the test passed, and then shortly after the result would change to a fail. I was able to capture this with screenshots once. These are screenshots from the same launch.
![result_changed](https://user-images.githubusercontent.com/22482646/174658783-c88e0d60-7a04-4648-a59f-02cefe2f7f7e.png)

2. I was able to capture logs. Github wouldn't let me upload html or xml files. I converted them to .txt files so hopefully you can convert them back successfully. the log should be .html and output .xml.

[log-20220620-131820 - Copy.txt](https://github.com/reportportal/agent-Python-RobotFramework/files/8942895/log-20220620-131820.-.Copy.txt)
[output-20220620-131820 - Copy.txt](https://github.com/reportportal/agent-Python-RobotFramework/files/8942897/output-20220620-131820.-.Copy.txt)

I was uploading the log file to Report Portal and thought Report Portal might be doing some post processing on the log, but even after disabling uploading the log file I still got the same results.

**Comment by Rickdickulous** [comment]
@HardNorth - Any update on this? I'm trying to understand if this is a Report Portal bug or a bug within the python agent. Thanks.

**Comment by Rickdickulous** [comment]
@HardNorth - I did extensive testing on this and can't make any sense of it. I set up a local report portal server and some test code to reproduce the issue and was able to with debug logging enabled.

For this run the Report Portal UI shows this as the failure:
![image](https://user-images.githubusercontent.com/22482646/176476432-1bb1378f-1b23-4778-9840-49cd8c3e052d.png)

Where as the log shows everything worked fine:
![image](https://user-images.githubusercontent.com/22482646/176498278-26543dce-2ff2-47b7-96bf-8a7bd88fae98.png)

Here's the log of this occurrence, you'll need to convert it back to html. Everything looks right from the report portal service and client side from what I can tell. Can you please confirm? Any suggestions on where to go/who to ask from here? Thanks again.

[attachment_118 - Copy.txt](https://github.com/reportportal/agent-Python-RobotFramework/files/9012773/attachment_118.-.Copy.txt)

**Comment by kpreety** [comment]
I am also observing same issue .

**Comment by HardNorth** [comment]
@Rickdickulous @kpreety Could you try using the latest version of the agent (5.3.0)? The only difference between this step and others is log event happened on "token check" step. If the case is in logging then 5.3.0 should help, since there are changes exactly in logging.

**Comment by Rickdickulous** [comment]
@HardNorth - The version numbers confused me a bit but I think I tested the scenario you intended. Here is a screenshot from my `pip list`.

![image](https://user-images.githubusercontent.com/22482646/177852087-29d4eec8-2cf9-42c8-985e-ae700909aaf0.png)


I am still able to create the skipped step marked as failed bug.
![image](https://user-images.githubusercontent.com/22482646/177852175-7f894a22-c91f-48fa-94fe-69986e0b1009.png)


Here is the associated log file.
[attachment_164.txt](https://github.com/reportportal/agent-Python-RobotFramework/files/9066587/attachment_164.txt)


Please let me know how I can assist.

**Comment by HardNorth** [comment]
@Rickdickulous The latest version is 5.3.0 as I said: [![PyPI](https://img.shields.io/pypi/v/robotframework-reportportal.svg?maxAge=259200)](https://pypi.python.org/pypi/robotframework-reportportal)

**Comment by HardNorth** [comment]
@Rickdickulous And I'm seeing the a batch log request in iteration 958:
![Screen Shot 2022-07-08 at 3 57 44 PM](https://user-images.githubusercontent.com/4364466/177996758-1dcf8688-f446-4d2b-82b8-bf8af343c571.png)

**Comment by Rickdickulous** [comment]
Ah glad I clarified. Here we go, latest versions of both. Tests are running now.

![image](https://user-images.githubusercontent.com/22482646/177987248-cde6ac5f-b3ad-4cb1-835f-9dab67e8ccf1.png)

I did notice the same thing with the bug occurring in the step that had the batch log request before. The batch logs go out at regular intervals so wasn't sure if that was a lead or not. 

I'll report back with test results later when I either see the bug again, or have run the scenario enough times to confidently say it's not occurring with the latest version of the agent.

**Comment by Rickdickulous** [comment]
@HardNorth - The bug is still present using the versions noted in my previous comment. Anecdotally it does seem to be happening less frequently. The rate of occurrence with the previous versions was 5/105,000, with this run being 1/100,000.

![image](https://user-images.githubusercontent.com/22482646/178010664-5944a9fb-4d53-4e45-908e-fc8a8eb1743b.png)

[attachment_315 - Copy.txt](https://github.com/reportportal/agent-Python-RobotFramework/files/9072772/attachment_315.-.Copy.txt)

**Comment by kpreety** [comment]
yeah, with latest version also, issue is still exists

**Comment by HardNorth** [comment]
No idea what might be the cause, sorry.

**Comment by Rickdickulous** [comment]
@HardNorth - Do you have any indication of the scope of this issue? Is it specific to the Robot Framework agent, the python report portal client, or the report portal application as a whole? I can't find anything wrong with the Robot Framework agent so suspect it's something higher up in the chain. Is this type of issue seen anywhere outside of Robot Framework use case?

**Comment by juhheikk** [comment]
I am also observing same issue.

**Comment by HardNorth** [comment]
@juhheikk I still need any reliable steps to reproduce.

**Comment by juhheikk** [comment]
We have over 800 testcases in a single output.xml (with reruns for some of the testcases) executed with pabot. After uploading this .xml file to report portal there comes a single testcase with Fail status, looking more deep it is not a real failure but the skipped steps are causing the test step (kw) to be reported as failure. And this does not come always to the same test, it is random behaviour.
![image](https://user-images.githubusercontent.com/92289862/235116155-733f811f-a62d-4a1f-9810-290f9a17d59d.png)
![image](https://user-images.githubusercontent.com/92289862/235117527-4e8bd2ec-9ace-47f4-8520-c3563e91edc0.png)

I tried to reproduce similar IF ENDs as in the above image, but I got just this, however there seems to be some "logic" error indicated by this one as well:

`*** Test Cases ***
First
  [Documentation]  first testcase
  Log  Use builtin logger

#Second
#  [Documentation]  second testcase
#  Fail  utterly

Third
  [Documentation]  third testcase
  ${var_x}=  Set Variable  ${FALSE}
  IF  ${TRUE}
    Log  In true IF
    IF  ${var_x}
      Log  IN IF
    ELSE
      Testing
    END
  END



*** Keywords ***
Testing
  [Arguments]  ${no_test}=${FALSE}  ${test}=${TRUE}
  IF  ${no_test}
    ${status}=  Run Keyword And Return Status  Fail  utterfly
    Log  error
  END
  IF  ${no_test}
    ${status}=  Run Keyword And Return Status  Fail  utterfly
    Log  error
  END
  IF  ${no_test}
    ${status}=  Run Keyword And Return Status  Fail  utterfly
    Log  error
  END
  IF  ${no_test}
    ${status}=  Run Keyword And Return Status  Fail  utterfly
    Log  error
  END
  IF  ${no_test}
    ${status}=  Run Keyword And Return Status  Fail  utterfly
    Log  error
  END
  IF  ${no_test}
    ${status}=  Run Keyword And Return Status  Fail  utterfly
    Log  error
  END
  IF  ${test}
    ${status}=  Run Keyword And Return Status  Log  test
    Log  testing
  END
`
`
$ pabot test.robot
$ python3 post_report.py -v RP_ENDPOINT:http://localhost:8080 -v RP_UUID:xx-xx-xx-xx-xx -v RP_PROJECT:local_demo -v RP_LAUNCH:superadmin_TEST_EXAMPLE -v RP_ATTACH_LOG:True -v RP_SKIPPED_ISSUE:False ./output.xml
`
None of the "Fail  utterly" are not even executed but those are shown as "Passed"
![image](https://user-images.githubusercontent.com/92289862/235119141-70b6ba36-5cdf-4ba7-add1-4b76e30cebb8.png)

**Comment by juhheikk** [comment]
Well I just thought that it could be closely related since it might not matter whether it is PASSED or FAILED that comes from the SKIPPED ones as there seems to be randomness involved. I can post the same info under another topic for you, no problem.

**Comment by HardNorth** [comment]
OK, let me take a look.

**Comment by Noordsestern** [comment]
We run in to the same:

We cannot reproduce it, but the pattern is this and happens at least once in each of our runs:
```robot
Some Keyword
    POST    ${some_url}     ${some_payload}
    IF    response.status_code >= 500
          Do Something
    END
```

I saw once that the IF statement was marked correctly as Skipped - but the POST was still marked as _In Progress_ for a few seconds. Once it got marked as _PASSED_ the IF switched its state to _FAILED_

We do use pabot to run Robot tests in parallel. But I would suppose ReportPortal can handle many messages.

**Comment by HardNorth** [comment]
@juhheikk Did it 30 times, no luck.
![Screenshot 2023-02-14 at 6 21 19 PM](https://github.com/reportportal/agent-Python-RobotFramework/assets/4364466/db56d0e8-5c6c-47be-84eb-f5c07e498f76)

**Comment by HardNorth** [comment]
@Noordsestern I would rather say this is pabot issue then, not our.

**Comment by HardNorth** [comment]
@Noordsestern Well, right now I don't really know. But your case actually makes sense, we use just a simple list to track item structure: https://github.com/reportportal/agent-Python-RobotFramework/blob/develop/robotframework_reportportal/listener.py#L51

@Rickdickulous 
If pabot (or any plugin / your test code) finishes the same item twice, or use threads to update keywords (I do see usage of threads in its code) it might cause collisions in item tracking. For threads it's because lists are not actually thread-safe in Python and do not provide any atomic way to update them. The only reason why it not happens often is GIL, Python switches between threads every 5 milliseconds and threads are not actually running in parallel. But when one thread stops somewhere between actual list item removing, and another thread start running and update the next item it gets the same id from the list and, therefore, mess with statuses.

Well, this agent is not actually written to handle such situations. And again, there is no our fault here, we just work synchronously.

**Comment by HardNorth** [comment]
Can someone test this with the agent version `5.4.0`?

**Comment by Rickdickulous** [comment]
I tested this fairly thoroughly last week and did not see this issue where skipped Robot Framework steps are reported as failures in Report Portal. Each test included a loop of 1,000 skipped steps. I ran this test 18 times, for a total of 18,000 skipped steps reported, with no erroneous failures in Report Portal. The one difference is that I got a new laptop that I'm running the tests on.

I will upate the package version in our test pipeline and report back here if I see any further cases of this bug. 

In the meantime I would consider this bug resolved as of version 5.4.0 as I am unable to reproduce it at this time.
